### PR TITLE
final version --MADE

### PR DESCRIPTION
**Pull request by IskornevME**
0. ### Архитектурные решения

- Структура проекта базируется на  cookiecutter-data-science.
- Данные загружаются с kaggle при помощи KaggleApi
- EDA отчет генерирует при помощи скрипта
- Для обработки данных написан кастомные трансформер с OneHotEncoder и StandardScaler
- При обучении данные делятся на train и test. Если при запуске train указать mdoel.grid_search=true,
   то запускается grid search и делается дополнительное разбиение на валидационную часть
- После обучения модель и тестовые данные сохраняются в отдельные пути (см. README)
- Для конфигов используется hydra (формат конфигов - yaml). Чтобы указать типы параметров используются   
   датаклассы. 
- Для обучения представлены 2 полноценные конфигурации: KNeighborsClassifier и RandomForestClassifier
- Для train и predict написаны отдельные функции, которые вызываются как утилиты командной строки
- Для загрузки данных и создания синтетических данных для тестов созданы специальные команды через entry
   points в setup.py
- Для логгирования используется logging
- Mlflow развернут локально. Зарегистрировано 2 модели (см. скрин ниже)
- В качестве метрик логгировал сразу несколько: f1-score, accuracy, roc_auc
- Для тестов используется unittest. Тесты проверяют преобразование данных, а также train и predict пайплайны
- Тесты и flake8 включены в CI, настроенный через GitHub Actions.

### Самооценка
1. В пулл-реквесте проведена самооценка, распишите по каждому пункту выполнен ли критерий или нет и на сколько баллов(частично или полностью) (**1 балл**) ☑️

2. Выполнено EDA, закоммитьте ноутбук в папку с ноутбуками (**1 балл**) Вы так же можете построить в ноутбуке прототип(если это вписывается в ваш стиль работы) ☑️

     Можете использовать не ноутбук, а скрипт, который сгенерит отчет, закоммитьте и скрипт и отчет (за это + **1   
     балл**) ☑️

3. Написана функция/класс для тренировки модели, вызов оформлен как утилита командной строки, записана в 
     readme инструкцию по запуску (**3 балла)** ☑️

4. Написана функция/класс predict (вызов оформлен как утилита командной строки), которая примет на вход 
     артефакт/ы от обучения, тестовую выборку (без меток) и запишет предикт по заданному пути, инструкция по  
     вызову записана в readme (**3 балла**) ☑️

5. Проект имеет модульную структуру (**2 балла**) ☑️

6. Использованы логгеры (**2 балла**) ☑️

7. Написаны тесты на отдельные модули и на прогон обучения и predict (**3 балла**) ☑️

8. Для тестов генерируются синтетические данные, приближенные к реальным (**2 балл**а) ☑️
     Используется DataSynthesizer

9. Обучение модели конфигурируется с помощью конфигов в json или yaml, закоммитьте как минимум 2 корректные 
     конфигурации, с помощью которых можно обучить модель (разные модели, стратегии split, preprocessing) (**3 
     балла**) ☑️

10. Используются датаклассы для сущностей из конфига, а не голые dict (**2 балла**) ☑️

11. Написан и протестирован кастомный трансформер(**3 балла**) ☑️

12. В проекте зафиксированы все зависимости (**1 балл**) ☑️

13. Настроен CI для прогона тестов, линтера на основе github actions (**3 балла**). ☑️

Дополнительные баллы=)

Используйте hydra для конфигурирования (https://hydra.cc/docs/intro/) - **3 балла** ☑️

Mlflow

разверните локально mlflow или на какой-нибудь виртуалке (**1 балл**) ☑️

залогируйте метрики (**1 балл**) ☑️

воспользуйтесь Model Registry для регистрации модели(**1 балл**) Приложите скриншот с вашим mlflow run DVC ☑️

выделите в своем проекте несколько entrypoints в виде консольных утилит (**1 балл**) ☑️

добавьте датасет под контроль версий (1 балл) ❌

сделайте dvc пайплайн(связывающий запуск нескольких entrypoints) для изготовления модели(1 балл) ❌

**Итого: 36 / 38 баллов**

Скриншоты mlflow:
image

Зарегистрированные модели:
image
image
image

### CODE REVIEW

File: ml_project/src/data/data_load.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import os
+from src.data import PATH_TO_DATA
```
Comment (SVizor42) [inline ml_project/src/data/data_load.py]: Локальные импорты идут после основных пакетов

File: ml_project/src/features/build_features.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+from typing import List
+# from sklearn.impute import SimpleImputer
+from sklearn.pipeline import Pipeline
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.preprocessing import OneHotEncoder, StandardScaler
+
+
+# CATEGORICAL_FEATURES = ['sex', 'cp', 'fbs', 'restecg',
+#                         'exang', 'slope', 'ca', 'thal']
+# NUMERIC_FEATURES = ['age', 'trestbps', 'chol', 'thalach', 'oldpeak']
```
Comment (SVizor42) [inline ml_project/src/features/build_features.py]: Лучше убрать закомментированные участки кода

File: ml_project/src/models/model_fit_predict.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+
+def train_model(
+        X_train_processed: pd.DataFrame, X_train: pd.DataFrame, target: pd.Series, param: TrainPipelineCfg
+) -> object:
+    with mlflow.start_run():
+        if param.model.name == 'knn':
+            model = KNeighborsClassifier(n_neighbors=param.model.num_n)
+            param_grid = {'n_neighbors': [3, 4, 5, 7, 9, 10],
+                          'metric': ['minkowski', 'euclidean']}
```
Comment (SVizor42) [inline ml_project/src/models/model_fit_predict.py]: Параметры лучше не хардкодить, а передавать через конфиг или хотя бы константу

File: ml_project/src/models/train_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+# logger.info("ssss")
+logger = logging.getLogger("train_model")
+
+
+@hydra.main(version_base=None, config_path='../config', config_name='config.yaml')
+def run_train_pipeline(param: TrainPipelineCfg):
+    logger.info(f'Start train with model {param.model.name}')
+    df = pd.read_csv(PATH_TO_DATA.joinpath('raw/heart_cleveland_upload.csv'))
+    X = df.drop(columns=['condition'])
+    target = df['condition']
```
Comment (SVizor42) [inline ml_project/src/models/train_pipeline.py]: Таргет лучше не хардкодить, а передавать через конфиг или константы

File: ml_project/reports/make_report.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import pandas as pd
+from io import StringIO
+import numpy as np
+from src.data import PATH_TO_DATA
```
Comment (SVizor42) [inline ml_project/reports/make_report.py]: Локальные импорты идут после основных пакетов

File: ml_project/src/models/train_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+# from entities import TrainConfig
+# logger.setLevel(logging.INFO)
+# formatter_stdout = logging.Formatter(
+#             "%(asctime)s\t%(levelname)s\t%(name)20s\t%(message)s",
+#             datefmt="%Y-%m-%d %H:%M:%S")
+# stream_handler = logging.StreamHandler()
+# stream_handler.setFormatter(formatter_stdout)
+# logger.addHandler(stream_handler)
+# logger.info(f"f1_score on train data = ")
+# logger.info("ssss")
```
Comment (SVizor42) [inline ml_project/src/models/train_pipeline.py]: Закомментированный блок лучше убрать

File: ml_project/src/models/train_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+logger = logging.getLogger("train_model")
+
+
+@hydra.main(version_base=None, config_path='../config', config_name='config.yaml')
+def run_train_pipeline(param: TrainPipelineCfg):
+    logger.info(f'Start train with model {param.model.name}')
+    df = pd.read_csv(PATH_TO_DATA.joinpath('raw/heart_cleveland_upload.csv'))
```
Comment (SVizor42) [inline ml_project/src/models/train_pipeline.py]: Лучше передавать пути ко всем файлам с данными и моделям через конфиг

File: ml_project/src/models/train_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        X_train,
+        param.features.categorical,
+        param.features.numerical
+    )
+    trans.fit()
+    X_train_processed = trans.transform()
+    y_train.reset_index(inplace=True, drop=True)
+    df_processed = pd.concat([X_train_processed, pd.DataFrame(y_train, columns=['condition'])], axis=1)
+    logger.info('Load process data to "data/processed/heart_cleveland_upload.csv"')
+    df_processed.to_csv(PATH_TO_DATA.joinpath('processed/heart_cleveland_upload.csv'), index=False)
```
Comment (SVizor42) [inline ml_project/src/models/train_pipeline.py]: Здесь обучение может "упасть", т.к. указанный путь к `processed` может не существовать

File: ml_project/src/models/model_fit_predict.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+import mlflow
+from random import shuffle
+from sklearn.model_selection import PredefinedSplit
+from sklearn.metrics import f1_score, accuracy_score, roc_auc_score
+from sklearn.model_selection import GridSearchCV
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.neighbors import KNeighborsClassifier
+from src.features import MyTransformer
+from src.entities import TrainPipelineCfg
+import os
```
Comment (SVizor42) [inline ml_project/src/models/model_fit_predict.py]: Локальные импорты идут после основных пакетов

File: ml_project/src/models/model_fit_predict.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+            )
+            trans.fit()
+            df_train_processed = trans.transform()
+            df_train_processed.set_index([pd.Index(train_index)], inplace=True)
+            trans = MyTransformer(
+                df_val,
+                param.features.categorical,
+                param.features.numerical
+            )
+            trans.fit()
```
Comment (SVizor42) [inline ml_project/src/models/model_fit_predict.py]: Зачем вновь обучать трансформер на валидационной части? Надо, например для скейлера, посчитывать статистики на тренировочной выборке и уже применять их к валидационной части

File: ml_project/src/models/predict_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+def run_predict_pipeline(
+        feature_test: pd.DataFrame, target_true: pd.Series, file_for_target: str, param: TrainPipelineCfg
+):
+    logger.info('Transform test features with our custom transformer')
+    trans = MyTransformer(
+        feature_test,
+        param.features.categorical,
+        param.features.numerical
+    )
```
Comment (SVizor42) [inline ml_project/src/models/predict_pipeline.py]: Трансформер не нужно вновь обучать. Его необходимо было сохранить в тренировочном модуле - либо отдельно, либо в виде готового пайплайна с моделью

File: ml_project/src/models/predict_pipeline.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+
+@click.command()
+@click.option('--path_train_data', '-ptd',
+              default=PATH_TO_DATA.joinpath('raw/features_test.csv'),
+              help='Please enter path to file with your train data. Default path - data/raw/features_test.csv')
+@click.option('--path_pred_target', '-ppt',
+              default=PATH_TO_DATA.joinpath("predicted_target.txt"),
+              help='Please enter path to file where you want save predicted target. Default path - '
+                   'data/predicted_target.txt')
```
Comment (SVizor42) [inline ml_project/src/models/predict_pipeline.py]: Почему не `@hydra`?

**Review by SVizor42 — CHANGES_REQUESTED** [review]
Посмотри, пожалуйста, комментарии. Нужно будет поправить некоторые вещи.

На что нужно обратить особое внимание:
1. Пути, имена признаков и таргетов нужно стараться передавать через конфиги.
2. Последовательность импортов.
3. Исключить из трекинга "мусорные" папки (`__pycahe__` и т.д.).
4. Добавить метки `hw1` и `MADE`/`TECHNOPARK`.

**Comment by IskornevME** [comment]
> 

Спасибо за замечания!
Постарался все поправить, посмотрите пожалуйста

File: ml_project/src/models/predict_pipeline.py (hunk 12 above)
Comment (IskornevME) [reply ml_project/src/models/predict_pipeline.py]: Показалось удобным, что можно передавать пути через флажок с помощью click. Хотя сейчас понимаю, что можно было обойтись гидрой и просто считывать аргументы из командной строки через argv

Comment (SVizor42) [reply ml_project/src/models/predict_pipeline.py]: Ок. Гидра тоже позволяет передавать параметры через аргументы командной строки :)
https://hydra.cc/docs/tutorials/basic/your_first_app/config_file/

File: ml_project/src/models/train_pipeline.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        param.features.categorical,
+        param.features.numerical
+    )
+    trans.fit(X_train)
+    with open(param.paths.path_to_transformer, 'wb') as f:
+        pickle.dump(trans, f)
+    X_train_processed = trans.transform(X_train)
+    y_train.reset_index(inplace=True, drop=True)
+    df_processed = pd.concat([X_train_processed, pd.DataFrame(y_train, columns=[param.features.target])], axis=1)
+    logger.info('Load process data to "data/processed/heart_cleveland_upload.csv"')
```
Comment (SVizor42) [inline ml_project/src/models/train_pipeline.py:43]: Путь можно также передать в строку через f-string

File: ml_project/src/models/train_pipeline.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
+import hydra
+import pickle
+from pathlib import Path
+import logging
+import os
+import pandas as pd
+from sklearn.model_selection import train_test_split
+from src.models import train_model
+from src.features import MyTransformer
+from hydra.core.config_store import ConfigStore
```
Comment (SVizor42) [inline ml_project/src/models/train_pipeline.py:10]: Гидру повыше локальных импортов

**Review by SVizor42 — APPROVED** [review]
Теперь уже лучше. Можно ветку сливать.

Рекомендую также почитать про `pytest` и фикстуры: https://docs.pytest.org/en/6.2.x/fixture.html

Я насчитал побольше баллов. **Сумма : 38 баллов**.
